proposals/NNNN-SPIRV-variable-address-space.md

```diff
@@ -0,0 +1,207 @@
+address spaces, making explicit the constraints some allocations have.
+Thus, it seems separating the address spaces for globals & locals would
+allow us to stay consistent with the rest of the language.
+
+## HLSL patterns to look for
+
+This section will explain why some HLSL patterns are hard to lower to SPIR-V.
+
+Note: HLSL does not implement references yet, but we have to make sure our
+design would allow us to implement them. For this reason, we'll assume HLSL
+has references.
+
+### Example 1:
+
+```hlsl
+static int a = 0;
+
+void foo() {
+  int b = 0;
+}
+```
+
+`a` and `b` both share the same address space. But on the SPIR-V side, `a`
+must be a `Private` variable, while `b` must be a `Function` variable.
+This requires the lowering pass to know the context of a variable.
+
+### Example 2:
+
+```hlsl
+static int a = 0;
+
+void foo() {
+  int& ref = a;
+  int b = ref;
+}
+```
+
+`a` is still `Private`, `b` still `Function`. But `ref` points to `a`.
+In SPIR-V, a variable cannot store a pointer pointing to another storage class.
+This means `ref` cannot be stored in a variable in the `Function` class.
+If `a` is `Private`, `ref` could only be declared as `Private`.
+
+### Example 3:
+
+```hlsl
+static int global = 0;
+
+int& foo(int& input, int select) {
+  return select ? input : global;
+}
+
+void main(int select) {
+  int local;
+  int& res1 = foo(local, select);
+  int& res2 = foo(global, select);
+}
+```
+
+`global` is still `Private`.
+`local` is `Function`.
+In SPIR-V, function declarations contains the return and parameters types,
+including the storage classes.
+This means, depending on the call-site, and the value of `select`, the
+return value and parameter would required either the `Function` or the
+`Private` storage class. When this selection depends on a runtime condition,
+this cannot be lowered to SPIR-V as-is.
+
+## Proposed solution: using 2 HLSL address spaces
+
+Thread-local, global variables will be put in the `hlsl_private` address
+space. Thread-local function-local variables will be put in the `default`
+address space.
+
+## Implementing the solution
+
+A new address space will be added to Clang: `hlsl_private`.
+This address space will be mapped to `Spirv::Private` on the SPIR-V backend,
+`PRIVATE_ADDRESS` for AMDGPU, and the address space `0` in DXIL.
+
+Clang codegen will add the new address space annotations, separating
+the `private` from the default.
+
+For the time being, the `private` address space will be marked as a subset of
+the `default` address space, allowing overload resolution for class methods:
+- an object in the `private` address space will be allowed to use a method
+  declared with a `this` in the default address space.
+
+Clang will emit an `addrspacecast` we will have to handle, but that's a known
+issue in address-space overload resolution, and not new to this proposal.
+
+## Alternative design considered
+
+### Force optimizations, and force inlining
```

Comment:
You mention this a bit above, but just to highlight: even with giving global and local variables their correct SPIRV address spaces, we're going to need to do some degree of this.

The member function problem is a source compatibility issue that I don't think we can just break. DXC even has some awful hacks to allow member functions to resolve on objects in cbuffers, despite the lack of proper constness on the implicit object (although that might be a small enough break that we can get away with not doing those hacks in Clang).